kore/test/Test/Kore/AllPath.hs

```diff
@@ -0,0 +1,347 @@
+
+emptyExecutionGraph :: AllPath.ProofState Integer -> ExecutionGraph
+emptyExecutionGraph = Strategy.emptyExecutionGraph
+
+insNode
+    :: (Gr.Node, AllPath.ProofState Integer)
+    -> ExecutionGraph
+    -> ExecutionGraph
+insNode = Strategy.insNode
+
+insEdge
+    :: (Gr.Node, Gr.Node)
+    -> ExecutionGraph
+    -> ExecutionGraph
+insEdge = Strategy.insEdge
+
+test_unprovenNodes :: [TestTree]
+test_unprovenNodes =
+    [ AllPath.unprovenNodes
+        (emptyExecutionGraph AllPath.Proven)
+        `satisfies_`
+        Foldable.null
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 1))
+        `satisfies_`
+        (not . Foldable.null)
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 1))
+        `equals`
+        (MultiOr.MultiOr [1])
+        $  "returns single unproven node"
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insNode (2, AllPath.Proven)
+        )
+        `equals_`
+        (MultiOr.MultiOr [0, 1])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Proven)
+            & insEdge (0, 2)
+        )
+        `equals_`
+        (MultiOr.MultiOr [1])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.Goal 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Goal 2)
+            & insEdge (1, 2)
+            & insNode (3, AllPath.Proven)
+            & insEdge (2, 3)
+        )
+        `equals_`
+        (MultiOr.MultiOr [])
+    , AllPath.unprovenNodes
+        (emptyExecutionGraph (AllPath.Goal 0)
+            & insNode (1, AllPath.GoalRem 1)
+            & insEdge (0, 1)
+            & insNode (2, AllPath.Proven)
+            & insEdge (0, 2)
+        )
+        `equals_`
+        (MultiOr.MultiOr [1])
+    ]
+
+-- | Simple program configurations for unit testing.
+data K = BorC | A | B | C | D | E | F | Bot
+    deriving (Eq, Ord, Show)
+
+matches :: K -> K -> Bool
+matches B BorC = True
+matches C BorC = True
+matches a b    = a == b
+
+instance EqualWithExplanation K where
+    compareWithExplanation = rawCompareWithExplanation
+    printWithExplanation = show
+
+type Goal = (K, K)
+
+type ProofState = AllPath.ProofState Goal
+
+type Rule = (K, K)
+
+ruleAC :: Rule
+ruleAC = (A, C)
+
+cyclicRule :: [Rule]
+cyclicRule = [(A, A)]
+
+ruleAB, ruleAF :: Rule
+ruleAB = (A, B)
+ruleAF = (A, F)
+
+ruleBC, ruleBD, ruleBF :: Rule
+ruleBC = (B, C)
+ruleBD = (B, D)
+ruleBF = (B, F)
+
+ruleCD, ruleCF :: Rule
+ruleCD = (C, D)
+ruleCF = (C, F)
+
+ruleDE :: Rule
+ruleDE = (D, E)
+
+ruleEF :: Rule
+ruleEF = (E, F)
+
+noRules :: [Rule]
+noRules = []
+
+oneRule :: [Rule]
+oneRule = [ruleAB]
+
+concurrentRules :: [Rule]
+concurrentRules = [ ruleAB, ruleAC ]
```

Comment:
I'd probably prefer to see [ (A, B), (A, C) ] used explicitly in code like

```haskell
proves    concurrentRules (A, BorC) "concurrent rules"
```

That is:

```haskell
proves    [(A, B),  (A, C)] (A, BorC) "concurrent rules"
``` 
... so that I only have to look one place to see *which* concurrent rules.  I might also use infix to make it more clear which are the rules and which is the goal:

```haskell
[(A, B),  (A, C)] `proves` (A, BorC) $ "concurrent rules"
``` 

But that's the English major in me again.